src/infra/StreamResponse.ts

```diff
@@ -0,0 +1,146 @@
+import { EnumOrLiteral, StreamEvent, StreamEventName } from "../models";
+import { RawResponse } from "../models/http/RawResponse";
+import { StreamEventCallback } from "../models/endpoints/stream/StreamEventCallback";
+
+/**
+ * Wrapper class around {@link RawResponse} that provides
+ * an interface for working with Chat's streaming data in
+ * browser and node environment.
+ *
+ * @public
+ */
+export class StreamResponse {
+  /**
+   * {@inheritdoc RawResponse}
+   *
+   * @public
+   */
+  readonly rawResponse: RawResponse;
+
+  private isConsumed = false;
+  private eventListeners: {
+    [E in StreamEventName]?: StreamEventCallback[];
+  } = {};
+
+  constructor(rawResponse: RawResponse) {
+    this.rawResponse = rawResponse;
+  }
+
+  /**
+   * Registers a function that will be called whenever the specified stream event occur.
+   *
+   * @public
+   *
+   * @param eventName - name of the event to listen
+   * @param cb - callback function to invoke when event occur
+   */
+  addEventListener(
+    eventName: EnumOrLiteral<StreamEventName>,
+    cb: StreamEventCallback
+  ) {
+    const cbs = this.eventListeners[eventName];
+    cbs ? cbs.push(cb) : (this.eventListeners[eventName] = [cb]);
+  }
+
+  /**
+   * Reads data from a stream response and invokes callbacks from event
+   * listeners for each chunk of data that is read.
+   *
+   * @remarks
+   * Once the data has been consumed from the stream, this method will
+   * simply return immediately on subsequent calls.
+   *
+   * @public
+   */
+  async consume(): Promise<void> {
+    if (this.isConsumed) {
+      console.error(
+        "Stream Error: data has been consumed from the stream. Cannot be read again."
+      );
+      return;
+    }
+    const resBody = this.rawResponse.body;
+    if (!resBody) {
+      return;
+    }
+    const streamCompleted =
+      "getReader" in resBody
```

Comment:
Should we indicate somewhere in the doc that the logic is somewhat split between Node and web? Not sure if that's something worth indicating to potential users or not. I saw you indicated this in the RawResponse doc, so might be overkill to indicate here as well. Up to you

Reply:
I think it's fine to omit it here since we are handling that split for user. The reason why I indicated this in RawResponse is because if user is accessing the raw response to do something themselves, they should be aware of the split in NodeJS vs WHATWG api as that changes what can be done to the response

Reply:
Ok, that sounds good!